firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/datagrams.txt

```diff
@@ -0,0 +1,122 @@
+      THEN("The parse status is equal to ok")
+      THEN("The payload of datagram changes after parse method is called") {
+      THEN("The length of the datagram is equal to the length of the payload from the input buffer")
+      THEN("The sequence of datagram is equal to sequence in the input buffer")
+      THEN("The input buffer is unchanged after parse")
+      THEN("The payload of the datagram is unchanged")
+
+    WHEN("An input buffer of 0 payload and length is parsed")
+      THEN("The parse status is equal to ok")
+      THEN("The payload of the datagram is equal to the initial payload")
+      THEN("The length of the datagram is equal to the length of the payload from the input buffer")
+      THEN("The sequence of datagram is equal to sequence in the input buffer")
+
+    WHEN("An input buffer of invalid length is parsed")
+      THEN("The parse status is equal to ok")
+      THEN("The payload of the datagram is equal to the initial payload")
+      THEN("The length of the datagram is equal to 0")
+      THEN("The sequence of datagram is equal to sequence in the input buffer")
+
+    WHEN("An input buffer of invalid sequence is parsed")
+      THEN("The parse status is equal to ok")
+      THEN("The payload of the datagram is equal to payload of the input buffer")
+      THEN("The length of the datagram is equal to the length in the input buffer")
+      THEN("The sequence of datagram is equal 0")
+
+/* Datagram Receiver
+
+Scenario: Datagram Receiver correctly parses datagrams into payloads
+  GIVEN("A Datagram receiver of buffer size 254 bytes")
+    WHEN("The input buffer is of invalid length")
+      THEN("The transform status is equal to invalid parse")
+
+    WHEN("The input buffer length is not equal to the actual length of the payload")
+      THEN("The transform status is equal to invalid parse")
+      THEN("The output datagram payload is equal to the input buffer paylaod")
+      THEN("The output datagram length is not equal to the input buffer payload length")
+
+    WHEN("The input buffer sequence is not equal to 0")
+      THEN("The transform status is equal to invalid parse")
+      THEN("The output datagram payload is equal to the input buffer paylaod")
+      THEN("The output datagram length is equal to the input buffer payload length")
+
+    WHEN("The input buffer has valid length, sequence and payload")
+      THEN("The transform status is equal to invalid parse")
+      THEN("The output datagram payload is equal to the input buffer paylaod")
+      THEN("The output datagram length is equal to the input buffer payload length")
+
+  // other tests similar to parse function
+
+/* Datagram sender
+
+Scenario: Datagram Sender correctly generates datagrams from payloads
+  GIVEN("A Datagram sender of buffer size 254 bytes")
+    WHEN("A valid payload is given as input to datagram sender")
+      THEN("The transform status is equal to ok")
+      THEN("The output datagram is as expected")
```

Comment:
We need to define tests about the behavior of incrementing and rollover of sequence numbers with DatagramSender